modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EDot.java

```diff
@@ -0,0 +1,288 @@
+import org.elasticsearch.painless.ir.ExpressionNode;
+import org.elasticsearch.painless.ir.ListSubShortcutNode;
+import org.elasticsearch.painless.ir.MapSubShortcutNode;
+import org.elasticsearch.painless.ir.NullSafeSubNode;
+import org.elasticsearch.painless.lookup.PainlessField;
+import org.elasticsearch.painless.lookup.PainlessLookupUtility;
+import org.elasticsearch.painless.lookup.PainlessMethod;
+import org.elasticsearch.painless.lookup.def;
+import org.elasticsearch.painless.symbol.ScriptRoot;
+
+import java.lang.reflect.Modifier;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.painless.lookup.PainlessLookupUtility.typeToCanonicalTypeName;
+
+/**
+ * Represents a field load/store and defers to a child subnode.
+ */
+public class EDot extends AExpression {
+
+    protected final boolean nullSafe;
+    protected final String value;
+
+    public EDot(Location location, AExpression prefix, boolean nullSafe, String value) {
+        super(location, prefix);
+
+        this.nullSafe = nullSafe;
+        this.value = Objects.requireNonNull(value);
+    }
+
+    @Override
+    Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input input) {
+        if (input.read == false && input.write == false) {
+            throw createError(new IllegalArgumentException("not a statement: result of dot operator [.] not used"));
+        }
+
+        Output output = new Output();
+
+        Output prefixOutput = prefix.analyze(classNode, scriptRoot, scope, new Input());
+
+        Class<?> targetType = prefixOutput.actual;
+        String targetCanonicalTypeName = PainlessLookupUtility.typeToCanonicalTypeName(targetType);
+
+        ExpressionNode expressionNode = null;
+
+        if (prefixOutput.actual.isArray()) {
+            if ("length".equals(value)) {
+                if (input.write) {
+                    throw createError(new IllegalArgumentException(
+                            "invalid assignment: cannot assign a value write to read-only field [length] for an array."));
+                }
+
+                output.actual = int.class;
+            } else {
+                throw createError(new IllegalArgumentException(
+                        "Field [" + value + "] does not exist for type [" + targetCanonicalTypeName + "]."));
+            }
+
+            DotSubArrayLengthNode dotSubArrayLengthNode = new DotSubArrayLengthNode();
+            dotSubArrayLengthNode.setLocation(location);
+            dotSubArrayLengthNode.setExpressionType(output.actual);
+            expressionNode = dotSubArrayLengthNode;
+        } else if (prefixOutput.actual == def.class) {
+            // TODO: remove ZonedDateTime exception when JodaCompatibleDateTime is removed
+            output.actual = input.expected == null || input.expected == ZonedDateTime.class || input.explicit ? def.class : input.expected;
```

Comment:
`input.expected == ZonedDateTime.class`?

Reply:
Under certain conditions we can optimize out a cast and allow an invokedynamic call to do the cast for us. These conditions are during an assignment (that is *not* a compound assignment), where the left expression resolves to a def type brace or dot. For something like `def x = new int[1]; x[0] = 1;` the generated called from the invokedynamic instruction can take in the 1 as an int type instead of having to go from int to Object and then Object to int if we were to leave in the cast to a def type. However, this only works for assignments where the types can be implicitly cast too from. Hence the exceptions here - if the user has done an explicit cast, if we do not know the target, and finally ZonedDateTime. ZonedDateTime is on this exception list because we allow unnatural implicit casts (ie invokedynamic doesn't do them natively) from JodaCompatibleDateTime to minimize disruptions to our users during our transition to Java time.